src/entry.rs

```diff
@@ -632,14 +660,86 @@ impl<'a> EntryFields<'a> {
                 })?;
             }
         }
-        if let Ok(mode) = self.header.mode() {
-            set_perms(dst, Some(&mut f), mode, self.preserve_permissions)?;
-        }
+        set_perms_ownerships(
+            dst,
+            Some(&mut f),
+            &self.header,
+            self.preserve_permissions,
+            self.preserve_ownerships,
+        )?;
         if self.unpack_xattrs {
             set_xattrs(self, dst)?;
         }
         return Ok(Unpacked::File(f));
 
+        fn set_ownerships(
+            dst: &Path,
+            f: &Option<&mut std::fs::File>,
+            uid: Option<u64>,
+            gid: Option<u64>,
+        ) -> Result<(), TarError> {
+            _set_ownerships(dst, f, uid, gid).map_err(|e| {
+                TarError::new(
+                    format!(
+                        "failed to set ownerships to uid={:?}, gid={:?} \
+                         for `{}`",
+                        uid,
+                        gid,
+                        dst.display()
+                    ),
+                    e,
+                )
+            })
+        }
+
+        #[cfg(unix)]
+        fn _set_ownerships(
+            dst: &Path,
+            f: &Option<&mut std::fs::File>,
+            uid: Option<u64>,
+            gid: Option<u64>,
+        ) -> io::Result<()> {
+            use std::os::unix::prelude::*;
+
+            // -1 (0xffffffff) means no ownership change in most libc implementations
```

Comment:
This comment about "most libc implementations" is a little worrisome, what about the "other" libc implementations? Does this simply not work in those situations?

(I'm not sure why uid/gid here are optional though)

Reply:
According to https://pubs.opengroup.org/onlinepubs/9699919799/functions/chown.html, `-1` is a POSIX-standard way to keep the current ID. There are probably some `libc` implementations that are not fully POSIX-compatible that do not understand this.

But yeah, if we want to handle the errors when reading UID/GID then this is irrelevant.